### PR TITLE
Display user friendly error message when importing from source tree

### PR DESCRIPTION
**Pull request by jjhelmus**
A user friendly message is displayed when an attempt to import skimage from the source tree is performed.  This message is not displayed if an in-place build has been complete.  In addition an error message is displayed if an incomplete build is detected. 

This functionality is implemented by checking for a stub module which is compiled during the build phase.  This feature is based off similar functionality from the scikit-learn package.

### CODE REVIEW

**Comment by jjhelmus** [comment]
Example error message when importing skimage from source directory:

```
$ python -c "import skimage"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "skimage/__init__.py", line 125, in <module>
    from . import __check_build
  File "skimage/__check_build/__init__.py", line 47, in <module>
    raise_build_error(e)
  File "skimage/__check_build/__init__.py", line 42, in raise_build_error
    %s""" % (e, local_dir, ''.join(dir_content).strip(), msg))
ImportError: No module named _check_build
___________________________________________________________________________
Contents of skimage/__check_build:
_check_build.pyx.md5      README                    __init__.pyc
__init__.py               _check_build.c            setup.pyc
_check_build.pyx          setup.py
___________________________________________________________________________
It seems that scikit-image has not been built correctly.

If you have installed scikit-image from source, please do not forget
to build the package before using it: run `python setup.py install` or
`make` in the source directory.

It appears that you are importing a local scikit-image source tree. For
this, you need to have an inplace install. Maybe you are in the source
directory and you need to try from another location.
```

**Comment by jjhelmus** [comment]
Merge conflict is due to skimage/__init__.py file which was modified in PR #1599.  Conflict can be resolved by hand without too much work.  Would a rebase be helpful?

**Comment by jjhelmus** [comment]
Also this PR addresses issue #1601

**Comment by stefanv** [comment]
A rebase would be helpful.  Also, I think since we are trying to make things easier for the user, we should avoid introducing confusion w.r.t. the check build module.  This should be invisible to users.

File: skimage/__check_build/__init__.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+""" Module to give helpful messages to the user that did not
+compile scikit-image properly.
+This checking code was stolen from scikit-learn & pyart.
```
Comment (stefanv) [inline skimage/__check_build/__init__.py]: Let's replace stolen by re-used or similar.

**Comment by stefanv** [comment]
Instead of importing a special submodule, why not use ``skimage._shared``, which is very likely to be imported either way.  Or, search for a known extension filename and make sure that something other than extension.py, extension.c, extension.md5 exists--this should catch 98% of cases without the extra complexity.

**Comment by jjhelmus** [comment]
I think we need to perform an import of a extension module not just check that a non .c, .py, .md5 file extension since the logic in the PR not only checks for imports in the source directory but also displays a helpful message when a corrupt or improper install is detected.  For example if a 64-bit extension is installed on a 32-bit system.

**Comment by jjhelmus** [comment]
New message when importing from the source tree:

```
$ python -c "import skimage"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "skimage/__init__.py", line 163, in <module>
    _raise_build_error(e)
  File "skimage/__init__.py", line 141, in _raise_build_error
    %s""" % (e, msg))
ImportError: cannot import name geometry
It seems that scikit-image has not been built correctly.

If you have installed scikit-image from source, please do not forget
to build the package before using it: run `python setup.py install` or
`make` in the source directory.

It appears that you are importing a local scikit-image source tree. For
this, you need to have an inplace install. Maybe you are in the source
directory and you need to try from another location.
```

**Comment by jjhelmus** [comment]
Where should this functionality be placed?  Currently it is in the `skimage/__init__.py` file which I'm not particularly pleased about.

**Comment by stefanv** [comment]
I'm a bit surprised that skimage gets imported during build.  Perhaps I shouldn't be.

This is getting close.  The two error messages, together, are still somewhat disjunctive.  Question: is it even possible for skimage to be installed without the binaries being compiled?  I guess this can only happen when running directly from source?

Let's say t is possible, then perhaps the displayed messages can follow something similar to the following logic:

- Are you trying to run from source?  If so, make sure you build.
- If you're not trying to run from source, your installation is broken and you need to re-install according to the instructions at -installation-url-from-docs-.

**Comment by jjhelmus** [comment]
Nearly all the subpackages import skimage._build at the top of their setup.py script which is why skimage is getting imported during build.

I do not think is is possible to install skimage without compiling the binaries.  It possible to install a version where the compiled binaries cannot be imported.  For example installing a wheel or conda package for the incorrect architecture (there are checks to prevent this but it can be done).  Also, having a shared library linked against a newer GLIB version than is available on the system (happens often with poorly built Linux conda packages).

The question is, do we need an overly helpful error message in these cases.  These errors should not be common, at least not as common as importing from the source tree, so I think a catch all "your install seems to be broken, re-install" should suffice.  

I'll look at the sklearn history for the file and see if they indicate why they have two messages.

**Comment by stefanv** [comment]
Thanks, that all makes sense and I agree with your assessment.

**Comment by jjhelmus** [comment]
Updated with language in the messages displayed to users, is http://scikit-image.org/docs/stable/install.html the best install instructions URL?

**Comment by stefanv** [comment]
This looks good to me!  Thanks, Jonathan.

**Pull request merged**